include/inl/gnc_utilities.inl

```diff
@@ -0,0 +1,138 @@
+//
+// include/gnc_utilities.hpp
+// PSim
+//
+// Contributors:
+//   Kyle Krol  kpk63@cornell.edu
+//
+// Pathfinder for Autonomous Navigation
+// Space Systems Design Studio
+// Cornell Univeristy
+//
+
+#include "../gnc_utilities.hpp"
+
+namespace gnc {
+namespace utl {
+
+template <typename T>
+constexpr void quat_conj(lin::Vector<T, 4> const &q, lin::Vector<T, 4> &res) {
+  res = { -q(0), -q(1), -q(2), q(3) };
+}
+
+template <typename T>
+constexpr void quat_conj(lin::Vector<T, 4> &q) {
+  quat_conj(q, q);
+}
+
+template <typename T>
+constexpr void quat_cross_mult(lin::Vector<T, 4> const &q1, lin::Vector<T, 4> const &q2,
+    lin::Vector<T, 4> &res) {
+  auto const qv1 = lin::ref<3, 1>(q1, 0, 0);
+  auto const qv2 = lin::ref<3, 1>(q2, 0, 0);
+  lin::ref<3, 1>(res, 0, 0) = lin::cross(qv2, qv1) + q1(3) * qv2 + q2(3) * qv1;
+  res(3) = q1(3) * q2(3) - lin::dot(qv1, qv2);
+}
+
+template <typename T>
+constexpr void rotate_frame(lin::Vector<T, 4> const &q, lin::Vector<T, 3> const &v, lin::Vector<T, 3> &res) {
+  auto const qv = lin::ref<3, 1>(q, 0, 0);
+  res = v + lin::cross(
+      (static_cast<T>(2.0) * qv).eval(),
```

Comment:
Could you explain how lin::cross works? Sometimes you are using `.eval()` on the returned value, but sometimes not. Whats the difference?

Reply:
So the `*.eval()` calls force the creation of a temporary object and prevents lazy evaluation. For example, writing `c = a + b` actually equates, in terms of types to `lin::Vector<> = lin::internal::Add<lin::Vector<>, lin::Vector<>>` and then a single for loop performs the addition and assignment all in one big swoop. In contrast writing `c = (a + b).eval()` gives us a temporary `lin::Vector<>` on the RHS and then a second for loop to perform assignment from the temporary to `c`.

This has the implication that writing `a += b` is essentially useless (and not supported by the library) because `a = a + b` results in the same computation.

Hopefully that makes sense. I apologize for the documentation being sparse.